r/src/array__to_vector.cpp

```diff
@@ -488,6 +488,47 @@ class Converter_Decimal : public Converter {
   }
 };
 
+class Converter_List : public Converter {
+ public:
+  explicit Converter_List(const ArrayVector& arrays) : Converter(arrays) {}
+
+  SEXP Allocate(R_xlen_t n) const { return Rcpp::List(no_init(n)); }
+
+  Status Ingest_all_nulls(SEXP data, R_xlen_t start, R_xlen_t n) const {
+    // nothing to do, list contain NULL by default
+    return Status::OK();
+  }
+
+  Status Ingest_some_nulls(SEXP data, const std::shared_ptr<arrow::Array>& array,
+                           R_xlen_t start, R_xlen_t n) const {
+    using internal::checked_cast;
+    auto list_array = checked_cast<arrow::ListArray*>(array.get());
+    auto values_array = list_array->values();
+
+    auto ingest_one = [&](R_xlen_t i) {
+      auto slice =
+          values_array->Slice(list_array->value_offset(i), list_array->value_length(i));
```

Comment:
`::Allocate()` makes an R list, and the various `::Ingest.*()` fill that list. 

I'll look if/how I can do this without actually calling `Slice()`, might be easy enough to just scan through `values_array` I guess.

Reply:
The challenging part is dealing with the bitmap and null_count.

Reply:
This can be a huge timesink, I think it's ok for now, but be wary that this will be a problem on large Array.

Reply:
I see. I don't think it's too much trouble to skip using Slice and use the bitmap of the `values_array` etc ... I'll have a look.

Reply:
🤔 yeah it's about the `Array__as_vector(slice)` which needs `slice` as a proper array already. The Converter api is designed to ingest all of an array, so I guess it would need some extra methods to only ingest a slice of an array without. I'll open another issue for that.

Reply:
https://issues.apache.org/jira/browse/ARROW-5628?filter=12344983